packages/services/src/view-service/fees.ts

```diff
@@ -60,9 +81,56 @@ export const extractAltFee = async (
   if (auctionWithdrawAsset) {
     const withdrawAuction = await indexedDb.getAuction(auctionWithdrawAsset);
     if (withdrawAuction.auction?.input?.assetId) {
-      return withdrawAuction.auction.input.assetId;
+      const inputAssetId = withdrawAuction.auction.input.assetId;
+      const assetId = await getAssetFromGasPriceTable(request, indexedDb, inputAssetId);
+      if (assetId) {
+        return assetId;
+      }
+    }
+  }
+
+  if (request.undelegations.length > 0) {
+    const assetId = await getAssetFromGasPriceTable(request, indexedDb);
+    if (assetId) {
+      return assetId;
+    }
+  }
+
+  throw new Error('Not able to pay fee for transaction');
+};
+
+// Attempts to find an alternative fee token for a transaction by checking the user's balance of
+// various gas price assets. It queries the IndexedDb for alternative gas prices and iterates
+// through them, returning the asset ID of the first asset for which the user has a balance.
+export const getAssetFromGasPriceTable = async (
```

Comment:
extracted all the complicated logic into a unified `getAssetFromGasPriceTable` function

Reply:
That's nice, I think we can make the choice of alternative fee assets even smarter if we filter out small balances when it's obvious that it won't be enough to pay the base tx fee (without considering the cost of actions).

Reply:
good point, thought about that early on, but wanted to limit the scope

Reply:
references https://github.com/penumbra-zone/web/issues/1595